cleanlab/cleanlab-studio
# Feature: CLI using data ingestion service and support for dataframe upload through Python API

### Description
This PR refactors the CLI to use our new data ingestion service. It also implements support for uploading text/tabular/external media datasets from a pandas DataFrame through the Python API and support for uploading ZIP datasets using a filepath through the Python API. 

See https://github.com/cleanlab/cleanlab-studio/pull/85 for more details on CLI changes

See https://www.notion.so/cleanlab/CLI-Python-API-c5df0da937f64d7d83e5c7888ace26b9 for mostly complete Python API design

A lot of changes in this file are just moving code around. Changes to pay attention to:
- `cleanlab_studio/internal/api/api.py`
- `cleanlab_studio/internal/dataset_source/*`
- `cleanlab_studio/internal/upload_helpers.py`
- `cleanlab_studio/internal/util.py`
- `cleanlab_studio/studio/studio.py`
- `cleanlab_studio/studio/upload.py`

TODO: 
- [x] fix test failures
- [x] update README to add instructions for setting upload base url
- [x] fix type checking

Still to come:
- [ ] schema inference method for Python API
- [x] schema overrides
- [x] PySpark support
- [x] ~~support for uploading from dataframe w/o specifying dataset name~~ error message if user uploads from dataframe without dataset name
- [ ] support for specifying metadata filepath/filepath column name for ZIP w/ metadata upload

### How to Test
Checkout backend branch `feat/upload-api-w-api-key`
CLI:
- Attempt to upload text/tabular and ZIP images datasets with and without user provided schema

Python API:
- Test Pandas support:
```
from cleanlab_studio import Studio
import pandas as pd
studio = Studio("{your_api_key}")
df = {load a dataset as pandas df}
dataset_id = studio.upload_dataset(df, {dataset_name})
```
- Check that dataset was successfully uploaded in local web app
- Test PySpark support using [this notebook](https://dbc-863d8aa2-05b3.cloud.databricks.com/?o=66949918647407#notebook/3150216172252497)
    - Check that dataset is successfully uploaded in hosted dev app
